analyzers/ImpactParameter.py

```diff
@@ -1,48 +1,233 @@
                         continue
-                    ptc.path.compute_IP(assumed_vertex)
-                    #jet.IP = ptc.path.IP
-            #setattr(event, self.cfg_ana.output, IPs)
+                    ptc.path.compute_IP(assumed_vertex,jet)
+                    
+                    ptc_IP_signif = 0
+                    if hasattr(ptc.path, 'points') == True and 'beampipe_in' in ptc.path.points:
+                        ptc.path.compute_theta_0()
+                        ptc.path.compute_IP_signif(ptc.path.IP,
+                                                   ptc.path.theta_0,
+                                                   ptc.path.points['beampipe_in'])
+                    else :
+                        ptc.path.compute_IP_signif(ptc.path.IP, None, None)
+                    
+                    if ptc.path.p4.Perp() > 1 and (ptc.path.IPx**2 + ptc.path.IPy**2)**0.5 < 0.002 and ptc.path.IPz**2 < 0.17**2 :
+                        
+                        dr = deltaR(jet.eta(), jet.phi(), ptc.eta(), ptc.phi())
+                        
+                        ptcs_pt.append(ptc.path.p4.Perp())
+                        ptcs_IP.append(ptc.path.IP)
+                        ptcs_IPxy.append((ptc.path.IPx**2 + ptc.path.IPy**2)**0.5)
+                        ptcs_IPz.append(ptc.path.IPz)
+                        ptcs_IP_signif.append((ptc.path.IP_signif))
+                        ptcs_IP_signif_sorted.append((ptc.path.IP_signif))
+                        x, y, z = ptc.path.coord_at_time(0)
+                        ptcs_x_at_time_0.append(x)
+                        ptcs_y_at_time_0.append(y)
+                        ptcs_z_at_time_0.append(z)
+                        ptcs_dr.append(dr)
+                        
+                        if self.tag_IP_b_LL:
+                            #import pdb; pdb.set_trace()
+                            ibin = self.ratio_IP.FindBin(ptc.path.IP)
+                            lhratio = self.ratio_IP.GetBinContent(ibin)
+                            if not lhratio == 0:
+                                ptc.path.IP_b_LL = math.log(lhratio)
+                                IP_b_LL += ptc.path.IP_b_LL
+                            if lhratio == 0:
+                                ptc.path.IP_b_LL = 0
+                        
+                        if self.tag_IPs_b_LL:
+                            #import pdb; pdb.set_trace()
+                            ibin = self.ratio_IPs.FindBin(ptc.path.IP_signif)
+                            lhratio = self.ratio_IPs.GetBinContent(ibin)
+                            if not lhratio == 0:
+                                ptc.path.IPs_b_LL = math.log(lhratio)
+                                IPs_b_LL += ptc.path.IPs_b_LL
+                            if lhratio == 0:
+                                ptc.path.IPs_b_LL = 0
+                            
+                ptcs_IP_signif_sorted.sort(reverse=True)
+            
+            if len(ptcs_IP_signif) < 2:
```

Comment:
I won't have a look at the following, please first reorganize as proposed above: 

- list of tuples 
- BTagInfo in ptc